### DISCUSSION TITLE
Add relation resolver using ExtendObjectType

### DISCUSSION THREAD

**Discussion by danrot** [ChilliCream/hotchocolate#3178] [Q&A] [answered]
Hey there :wave: I am just wondering if the documentation at https://chillicream.com/docs/hotchocolate/api-reference/object-type/ is correct. I have a Query class that returns a list of Employees:

```csharp
// Removed
```

The Employee class has a one-to-many relationship to the Absence, which should be exposed via GraphQL as well. I was wondering what’s the easiest possible way to do that. The way I’ve understood it, would be to introduce an object type extension using “Pure Code-First” as mentioned in the link above, so that’s what I’ve came up with:

```csharp
// Removed
```

My first guess was that I should be able to introduce that GetAbsences function in the Query class as well, and the Parent Attribute would automatically recognize when to use that function, but that does not seem to work, so I switched to the approach with the ExtendObjectType attribute. I also registered that class in the Startup.cs file of ASP.NET:

```csharp
// Removed
```

But when I try to query absences from the employees now, I always get an empty list, and there are no logs indicating that a database request to that table was done. A RTFM with a link would be totally fine, but I couldn’t find any V11 documentation covering this in more detail than the one above :confused:

**Comment by gojanpaolo** [comment, marked as answer]
@danrot Can you create a project that reproduces the issue?

I tried running the code you provided but it's working fine.

See working sample project here: https://github.com/gojanpaolo/hotchocolate-3178

</details>

Runnning the following query:

```graphql
query {
  employee(id: 1) {
    absences {
      id
    }
  }
}
```

... returns the following response

```json
{
  "data": {
    "employee": {
      "absences": [
        {
          "id": 1
        }
      ]
    }
  }
}
```

**Reply by michaelstaib** [reply]
The code actually looks good. Maybe @danrot can also post his query.

**Reply by gojanpaolo** [reply]
I tested further and got the following error when getting `absences` of multiple `employees`.

```graphql
query {
  employees {
    nodes {
      absences {
        id
      }
    }
  }
}
```

```
...
"message": "A second operation was started on this context before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext. For more information on how to avoid threading issues with DbContext, see https://go.microsoft.com/fwlink/?linkid=2097913.",
"stackTrace": "   at Microsoft.EntityFrameworkCore.Internal.ConcurrencyDetector.EnterCriticalSection()\r\n   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.Enumerator.MoveNext()\r\n   at HotChocolate.Execution.Processing.ResolverTask.<>c__DisplayClass13_0.<ExecuteResolverPipelineAsync>b__0()\r\n   at System.Threading.Tasks.Task`1.InnerInvoke()\r\n   at System.Threading.Tasks.Task.<>c.<.cctor>b__277_0(Object obj)\r\n   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)\r\n--- End of stack trace from previous location ---\r\n   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)\r\n   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)\r\n--- End of stack trace from previous location ---\r\n   at HotChocolate.Execution.Processing.ResolverTask.ExecuteResolverPipelineAsync(CancellationToken cancellationToken)\r\n   at HotChocolate.Execution.Processing.ResolverTask.TryExecuteAsync(CancellationToken cancellationToken)"
...
```

This should be solved by using the `DbContextFactory` as shown in [graphql-workshop Session 3](https://github.com/ChilliCream/graphql-workshop/blob/master/docs/3-understanding-dataLoader.md)

**Reply by michaelstaib** [reply]
Well, that is because you are using a single DBContext :)

you need to use pooled db contexts at least with version 11.0 and 11.1. We will teach the execution engine about services that can not be used in parallel with the May release.

**Reply by gojanpaolo** [reply]
@michaelstaib Yes, 
@danrot I recommend going over the [graphql-workshop](https://github.com/ChilliCream/graphql-workshop) if you haven't done so already. IMO the workshop is a great resource to start learning HC and GraphQL. :)

**Reply by danrot** [reply]
Hey, I've had a quick look at the repo from @gojanpaolo, and it indeed looks similar to what I've been doing... I am using the following query:

```gql
// Removed
```

And the result looks like this:

```json
// Removed
```

I'll have another closer look soonish and will report back.

**Reply by danrot** [reply]
Talked to @michaelstaib and he found the issue 🙂 The `ExtendObjectType`(currently) cannot override existing properties, and my `Employee` class already has a `Absences` property defined. The easiest way to solve the issue was to just use [`UseProjection`](https://chillicream.com/docs/hotchocolate/fetching-data/projections/), which made the entire `EmployeeResolvers` class obsolete.

PS: @gojanpaolo it would be nice if you could delete the repository you have created again. I am using that example in a classroom situation, and I don't want the students to find an almost finished example easily on the web 🙂

**Reply by michaelstaib** [reply]
We will change this behavior with 11.1

**Reply by ThisNoName** [reply]
> Well, that is because you are using a single DBContext :)
> 
> you need to use pooled db contexts at least with version 11.0 and 11.1. We will teach the execution engine about services that can not be used in parallel with the May release.

Will the parallel setting be keyed/grouped or something?  Like when business layer talking to different databases,  it makes sense to still allow parallel between different databases.  Thanks
